### PR TITLE
SCRUM-166, 26, 202, 203 -- User Customization, Profile Revamp, Dashboard Revamp, Mock Test Revamp

### PR DESCRIPTION
**Pull request by dahlia-factory**
This PR addresses [SCRUM-166](https://knightwise.atlassian.net/jira/software/projects/SCRUM/boards/1?selectedIssue=SCRUM-166), [SCRUM-26](https://knightwise.atlassian.net/jira/software/projects/SCRUM/boards/1?selectedIssue=SCRUM-26), [SCRUM-202](https://knightwise.atlassian.net/jira/software/projects/SCRUM/boards/1?selectedIssue=SCRUM-202), and [SCRUM-203](https://knightwise.atlassian.net/jira/software/projects/SCRUM/boards/1?selectedIssue=SCRUM-166)

- This PR adds and allows for the customization of user profiles with the planned store feature, including profile pictures, flairs, and backgrounds.  
image

- This PR adds an overhauled Dashboard to better reflect current site design and goals
image

- Mock Tests are now customizable, and the new menu allows you to choose subject matter, set time limits, and number of questions
image

- This PR fixes an issue regarding questions with the category "Input/Output' being rendered, and does some mapping so the title shows up correctly.
- Slightly tightened page margins so that content is less boxed in by whitespace

### CODE REVIEW

**Comment by daniellandsman1** [comment]
This will require some _very careful_ merge conflict resolution to make sure the changes in #32 aren't lost, I can help with that.

**Review by minervatbd — APPROVED** [review]

**Comment by daniellandsman1** [comment]
See commit [20f9c73](https://github.com/KnightWiseUCF/knightwise/pull/33/commits/20f9c7399e5d75f3c1fcc3f58e38bfe80c3e4af2) for how I resolved the conflicts.

File: frontend/src/pages/ProfilePage.tsx [hunk 1]
```diff
@@ -0,0 +1,742 @@
...
+  };
+
+  const activeUser = isUsernameProfileRoute
+    ? null
+    : isReadOnlyProfile
+      ? publicProfile?.user ?? null
+      : isInvalidProfileRoute
+        ? null
+        : user;
+  const activeEquippedItems = isUsernameProfileRoute
```
Comment (daniellandsman1) [inline frontend/src/pages/ProfilePage.tsx:290]: NON-BLOCKER:
This causes linter warnings. I don't think just warnings will fail the pipeline, so this doesn't even need to be fixed, it's just something I noticed.

**Comment by daniellandsman1** [comment]
NON-BLOCKER:
Edge case worth noting:
- Building a mock test with a greater question count than the number of questions available for the selected topics (think selecting 50 questions and just picking Stacks) will cap the actual number of questions served as all the questions from the topic. This is fine and doesn't break anything, but the user may wonder why they only got 2 questions instead of the 50 they selected.
  - In the future it could be worth displaying a message like "Stacks doesn't have 50 questions, displaying all available questions instead" if question count > number of questions available for a topic, or even preventing a user from incrementing question count above the sum of the available questions selected.
  - Either way this could be annoying to implement and our handling of the edge case is good enough as is so we shouldn't worry about this for now. Just something to keep in mind.

File: frontend/src/components/ProfessorDashboard.tsx [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+
+const topicCategoryMap: Record<string, string[]> = {
+  "Introductory Programming": ["Input/Output", "Branching", "Loops", "Variables"],
+  "Simple Data Structures": ["Arrays", "Linked Lists", "Strings"],
+  "Object Oriented Programming": ["Classes", "Methods"],
+  "Intermediate Data Structures": ["Trees", "Stacks"],
+  "Complex Data Structures": ["Heaps", "Tries"],
+  "Intermediate Programming": ["Bitwise Operators", "Dynamic Memory", "Algorithm Analysis", "Recursion", "Sorting"],
+};
+
```
Comment (daniellandsman1) [inline frontend/src/components/ProfessorDashboard.tsx:33]: Part of my batch of suggested changes. Using **GET** `/api/admin/published` eliminates the need for this map.

```suggestion
```

File: frontend/src/components/ProfessorDashboard.tsx [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+      let parsedUserId: number | null = null;
+      try {
+        const rawUserData = localStorage.getItem("user_data");
+        const parsedUser = rawUserData ? JSON.parse(rawUserData) as { id?: number; ID?: number } : {};
+        const value = Number(parsedUser.id ?? parsedUser.ID);
+        parsedUserId = Number.isFinite(value) ? value : null;
+      } catch {
+        parsedUserId = null;
+      }
+
```
Comment (daniellandsman1) [inline frontend/src/components/ProfessorDashboard.tsx:101]: Part of my batch of suggested changes. Ownership filtering is already done on the backend by **GET** `/api/admin/published`, so this is no longer necessary.


```suggestion
```

File: frontend/src/components/ProfessorDashboard.tsx [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+      const draftPromise = api.get<DraftListResponse>("/api/admin/drafts");
+
+      const publishedPromise = Promise.allSettled(
+        Array.from(new Set(defaultSubcategories)).map((subcategory) =>
+          api.get<RawQuestion[]>(`/api/test/topic/${encodeURIComponent(subcategory)}`)
+        )
+      );
```
Comment (daniellandsman1) [inline frontend/src/components/ProfessorDashboard.tsx:108]: Part of my batch of suggested changes. This is the main change, using the `/published` endpoint instead of iterating through every topic and getting incorrect programming question info.

```suggestion
      const publishedPromise = api.get<{ published: RawQuestion[] }>("/api/admin/published");
```

File: frontend/src/components/ProfessorDashboard.tsx [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+            id: question.ID,
+            title: String(question.QUESTION_TEXT || `Question #${question.ID}`),
+            category: String(question.CATEGORY || "General"),
+            subcategory: String(question.SUBCATEGORY || "General"),
+            status: "Published",
+          });
+        });
+      } else {
+        errors.push("Failed to load published questions.");
+      }
```
Comment (daniellandsman1) [inline frontend/src/components/ProfessorDashboard.tsx:163]: Part of my batch of suggested changes. We don't need to merge and dedup here since `/published` returns a single response with a single array.

```suggestion
      if (publishedResult.status === "fulfilled") {
        const publishedRows = Array.isArray(publishedResult.value.data?.published)
          ? publishedResult.value.data.published
          : [];

        publishedRows.forEach((question) => {
          if (typeof question.ID === "number") {
            collected.push({
              id: question.ID,
              title: String(question.QUESTION_TEXT || `Question #${question.ID}`),
              category: String(question.CATEGORY || "General"),
              subcategory: String(question.SUBCATEGORY || "General"),
              status: "Published",
            });
          }
        });
      } else {
        errors.push("Failed to load published questions.");
      }
```

**Review by daniellandsman1 — APPROVED** [review]
The frontend looks great, this is the best-looking part of our app yet!

NON-BLOCKER:
`ProfessorDashboard.tsx` calls **GET** `/api/test/topic` instead of **GET** `/api/admin/published` to get published questions. I'd recommend matching the structure of `ProfessorDraftsPage.tsx`, which utilizes the `/published` endpoint. 

The only reason this is a non-blocker is because professors can't create programming questions right now. If we agree that this is something that will _never_ change, this won't break anything. But I'd still recommend making the changes for the following reasons:

1. Performance: The current implementation makes almost 20 API calls by iterating through every topic, when only one `/published` call is needed.

2. Consistency: The suggested changes is pretty much exactly how `ProfessorDraftsPage.tsx` deals with getting published questions for a professor.

3. Endpoint purpose: `/topic` is for topic practice session generation, not question enumeration. The logic was changed in #32 to have the endpoint return a maximum of one programming question to respect programming question submission limits. Again, won't cause an issue right now since professors can't create programming questions but this is just a case of an endpoint being used in a way that doesn't align with its purpose. `/published` was created specifically for scenarios like this where a professor's published questions need to be fetched.

Note that the suggested changes all go hand in hand, feel free to either apply them all or disregard them all; I would avoid applying them partially.

**Comment by daniellandsman1** [comment]
Merging this in, but creating [SCRUM-208](https://knightwise.atlassian.net/browse/SCRUM-208) to address the issues I mentioned.
